File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/UnityDEBUG/ForDebug/Assets/Scripts/CharacterDriven/BaseKalmanFilter.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public abstract class BaseKalmanFilter<TState, TMeasurement>
{
    // Fields
    private Func<TMeasurement, float[]> measurementConvertFunc;
    protected float[] state;
    private Func<float[], TState> stateConvertBackFunc;
    private Func<TState, float[]> stateConvertFunc;

    // Methods
    protected BaseKalmanFilter(TState initialState, float[,] initialStateError, int measurementVectorDimension, int controlVectorDimension, Func<TState, float[]> stateConvertFunc, Func<float[], TState> stateConvertBackFunc, Func<TMeasurement, float[]> measurementConvertFunc)
    {
        this.initalize(initialState, initialStateError, measurementVectorDimension, controlVectorDimension, stateConvertFunc, stateConvertBackFunc, measurementConvertFunc);
    }

    internal float[] CalculateDelta(float[] measurement)
    {
        float[] b = this.MeasurementMatrix.Multiply(this.state);
        return measurement.Subtract(b, false);
    }

    public void Correct(TMeasurement measurement)
    {
        this.CorrectInternal(this.measurementConvertFunc(measurement));
    }

    protected abstract void CorrectInternal(float[] measurement);
    private void initalize(TState initialState, float[,] initialStateError, int measurementVectorDimension, int controlVectorDimension, Func<TState, float[]> stateConvertFunc, Func<float[], TState> stateConvertBackFunc, Func<TMeasurement, float[]> measurementConvertFunc)
    {
        float[] numArray = stateConvertFunc(initialState);
        this.StateVectorDimension = numArray.Length;
        this.MeasurementVectorDimension = measurementVectorDimension;
        this.ControlVectorDimension = controlVectorDimension;
        this.state = numArray;
        this.EstimateCovariance = initialStateError;
        this.stateConvertFunc = stateConvertFunc;
        this.stateConvertBackFunc = stateConvertBackFunc;
        this.measurementConvertFunc = measurementConvertFunc;
    }

    public void Predict()
    {
        this.Predict(null);
    }

    public void Predict(float[] controlVector)
    {
        this.predictInternal(controlVector);
    }

    protected abstract void predictInternal(float[] controlVector);

    // Properties
    public float[,] ControlMatrix { get; set; }

    public int ControlVectorDimension { get; private set; }

    public float[,] EstimateCovariance { get; protected set; }

    public float[,] KalmanGain { get; protected set; }

    public float[,] MeasurementMatrix { get; set; }

    public float[,] MeasurementNoise { get; set; }

    public int MeasurementVectorDimension { get; private set; }

    public float[,] ProcessNoise { get; set; }

    public float[,] ResidualCovariance { get; protected set; }

    public float[,] ResidualCovarianceInv { get; protected set; }

    public TState State
    {
        get
        {
            return this.stateConvertBackFunc(this.state);
        }
    }

    public int StateVectorDimension { get; private set; }

    public float[,] TransitionMatrix { get; set; }
}

File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/KeyPointCenter/KeyPoint/BaseKalmanFilter.swift
//
//  BaseKalmanFilter.swift
//  MetaScripts
//
//  Created by ZhaoTianyu on 2024/7/8.
//

import Foundation

public class BaseKalmanFilter<TState, TMeasurement> {
    typealias MeasurementConvertFunc = (TMeasurement) -> [Float]
    typealias StateConvertBackFunc = ([Float]) -> TState
    typealias StateConvertFunc = (TState) -> [Float]
    
    var measurementConvertFunc: MeasurementConvertFunc?
    var state: [Float]?
    var stateConvertBackFunc: StateConvertBackFunc?
    var stateConvertFunc: StateConvertFunc?
    var controlMatrix: [[Float]]?
    var controlVectorDimension: Int?
    var estimateCovariance: [[Float]]?
    var kalmanGain: [[Float]]?
    var measurementMatrix: [[Float]]?
    var measurementNoise: [[Float]]?
    var measurementVectorDimension: Int?
    var processNoise: [[Float]]?
    var residualCovariance: [[Float]]?
    var residualCovarianceInv: [[Float]]?
    var State: TState? {
        get {
            guard let state = state else {
                return nil
            }
            return stateConvertBackFunc?(state)
        }
    }
    var stateVectorDimension: Int?
    var transitionMatrix: [[Float]]?
    
    required init(
        initialState: TState,
        initialStateError: [[Float]],
        measurementVectorDimension: Int,
        controlVectorDimension: Int,
        stateConvertFunc: @escaping StateConvertFunc,
        stateConvertBackFunc: @escaping StateConvertBackFunc,
        measurementConvertFunc: @escaping MeasurementConvertFunc
    ) {
        let numArray = stateConvertFunc(initialState)
        self.stateVectorDimension = numArray.count
        self.measurementVectorDimension = measurementVectorDimension;
        self.controlVectorDimension = controlVectorDimension;
        self.state = numArray;
        self.estimateCovariance = initialStateError;
        self.stateConvertFunc = stateConvertFunc
        self.stateConvertBackFunc = stateConvertBackFunc
        self.measurementConvertFunc = measurementConvertFunc
    }
    
    func calculateDelta(measurement: [Float]) -> [Float] {
        guard let measurementMatrix = measurementMatrix, let state = state else {
            return []
        }
        
        let b = VectorF.mul(measurementMatrix, state)
        return VectorF.sub(measurement, b)
    }
    func correct(measurement: TMeasurement) {
        guard let measurement = measurementConvertFunc?(measurement) else {
            return
        }
        correctInternal(measurement: measurement)
    }
    func correctInternal(measurement: [Float]) {
        
    }
    
    func predict() {
        predict(controlVector: nil)
    }
    func predict(controlVector: [Float]?) {
        predictInternal(controlVector: controlVector)
    }
    func predictInternal(controlVector: [Float]?) {
        
    }
}
